modules/gapi/include/opencv2/gapi/s11n.hpp

```diff
@@ -175,16 +189,130 @@ IIStream& operator>> (IIStream& is, std::vector<T> &ts) {
+    static auto try_call_serialize(gapi::s11n::IOStream& os, const util::any& arg, int)
+        -> sfinae_true<decltype(S11N<Q>::serialize(os, util::any_cast<Q>(arg)), void())> {
+
+        S11N<Q>::serialize(os, util::any_cast<Q>(arg));
+        return sfinae_true<void>{};
+    }
+
+    // FIXME: Add compile-time check that this can't be called for custom types.
+    //        Such behaviour can alternatively be implemented in S11N<T> with std::enable_if
+    //        contraining T to be types defined by framework.
+    template<typename Q = T>
+    static auto try_call_serialize(gapi::s11n::IOStream& os, const util::any& arg, long)
+        -> sfinae_true<decltype(os << std::declval<const typename std::add_lvalue_reference<Q>::type>(), void())> {
+
+        os << util::any_cast<Q>(arg);
+        return sfinae_true<void>{};
+    }
+
+    template<typename Q = T>
+    static std::false_type try_call_serialize(gapi::s11n::IOStream &os, const util::any& arg, ...);
+
+    static void call_serialize(gapi::s11n::IOStream& os, const util::any& arg) {
+        try_call_serialize<T>(os, arg, 0);
+    }
+};
+
+template<typename T>
+std::function<void(gapi::s11n::IOStream& os, const util::any& arg)>
+wrap_serialize<T, cv::GCompileArg>::serialize =
+        decltype(try_call_serialize(
+                    std::declval<typename std::add_lvalue_reference<gapi::s11n::IOStream>::type>(),
+                    std::declval<typename std::add_lvalue_reference<util::any>::type>(),
+                    int()))::value
+        ? &call_serialize : nullptr;
+
+
+template<typename T> struct wrap_deserialize
+{
+private:
+    template<typename Q = T>
+    static auto call_deserialize(gapi::s11n::IIStream& is, int)
+        -> decltype(S11N<Q>::deserialize(is)) {
+        return S11N<Q>::deserialize(is);
+    }
+
+    // FIXME: Add compile-time check that this can't be called for custom types.
+    //        Such behaviour can alternatively be implemented in S11N<T> with std::enable_if
+    //        contraining T to be types defined by framework.
+    template<typename Q = T>
+    static auto call_deserialize(gapi::s11n::IIStream& is, long)
+        -> decltype(is >> std::declval<typename std::add_lvalue_reference<Q>::type>(),
+                    Q()) {
+        Q obj;
+        is >> obj;
+        return obj;
+    }
+
+    template<typename Q = T>
+    static cv::GCompileArg call_deserialize(gapi::s11n::IIStream&, ...) {
+        return GCompileArg { };
+    }
+
+public:
+    static auto deserialize(gapi::s11n::IIStream& is)
+        -> decltype(call_deserialize<T>(is, 0)) {
+        return call_deserialize<T>(is, 0);
+    }
+};
 } // namespace detail
 } // namespace s11n
+
+namespace detail
+{
+template<typename T>
+GAPI_EXPORTS GCompileArg deserialize_arg(cv::gapi::s11n::IIStream& is, const std::string& tag) {
+    if (tag == cv::detail::CompileArgTag<T>::tag()) {
+        return GCompileArg { cv::gapi::s11n::detail::wrap_deserialize<T>::deserialize(is) };
```

Comment:
I'd expect you to call a callback from `S11N` directly here. Why one more level of indirection? You know your type already.

Reply:
Removed extra indirection level.